pkg/itemblock/actions/pod_action.go

```diff
@@ -0,0 +1,63 @@
+/*
+Copyright the Velero contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package actions
+
+import (
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	corev1api "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/vmware-tanzu/velero/pkg/actionhelpers"
+	v1 "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	"github.com/vmware-tanzu/velero/pkg/plugin/velero"
+)
+
+// PodAction implements ItemBlockAction.
+type PodAction struct {
+	log logrus.FieldLogger
+}
+
+// NewPodAction creates a new ItemBlockAction for pods.
+func NewPodAction(logger logrus.FieldLogger) *PodAction {
+	return &PodAction{log: logger}
+}
+
+// AppliesTo returns a ResourceSelector that applies only to pods.
+func (a *PodAction) AppliesTo() (velero.ResourceSelector, error) {
+	return velero.ResourceSelector{
+		IncludedResources: []string{"pods"},
+	}, nil
+}
+
+// GetRelatedItems scans the pod's spec.volumes for persistentVolumeClaim volumes and returns a
+// ResourceIdentifier list containing references to all of the persistentVolumeClaim volumes used by
+// the pod. This ensures that when a pod is backed up, all referenced PVCs are backed up along with the pod.
+func (a *PodAction) GetRelatedItems(item runtime.Unstructured, backup *v1.Backup) ([]velero.ResourceIdentifier, error) {
```

Comment:
_Sorry if we have discussed this in the design and if so just remind me again:_

What is the consequence when `GetRelatedItems` and `Execute` returns differently for the same object (e.g., pod, PVC)? This is probably since the are called at different time

Reply:
We did discuss in the design, but I can answer here as well. In many cases, the lists will be the same. However, there are some cases where they'll differ:
1) Items that don't exist yet at item collection won't be returned by the new IBA plugins but may be returned from BIA plugins (i.e. VolumeSnapshots created during CSI PVC action) -- these will not be included in the item block list, but they will be backed up just like now right when the BIA plugin returns the additional item).
2) Items may be returned by IBA plugins but don't currently get returned by any BIA plugins. The only example of this right now is the new functionality in this PR to return pods that are mounted by a PVC -- to handle making sure that shared RWX volumes have their pod backup coordinated. In this case, the ItemBlock will contain these extra pods -- no BIA plugins will return them, but that's unnecessary since the pods will be backed up in the order they are listed in the ItemBlock (with hooks coordinated to run before and after the block, of course).

In the case where the same item is returned by both (i.e. the PVC for a pod), the PVC will be included in the item block, but since the BIA returns the PVC as an additional item (with the API meaning of "back up this additional item right now, then continue with the current item), when the ItemBlock processing gets to that PVC later in the block, it's already in the BackedUpItems map, so velero will move on to the next item in the block.

Reply:
> The only time this makes any difference at all would be for a plugin which needs to make APIServer calls to the cluster itself to identify related items

Yes, so I once just thought about this --- is it a good idea to have the plugins to retrieve objects from the same cache created during the item collection? In that way, we have an almost consistent view across Velero backup and plugins.
But for sure that involves many changes. And it is also out of the scope of this itemblock implementation.

Reply:
As you said, that would involve many changes. We'd need to expose the item collector list of items to plugins somehow. Not only is this not part of the plugin API right now, but since plugins run in a separate process, I don't really see an easy way of doing this. Also, there's no guarantee that the resources that plugins may need to query are even in the object list for the backup.